File: macos/Turkify/Turkify/ModelDiscovery.swift
import Foundation

/// Keşfedilen bir yerel LLM sunucusu ve modelleri (tür/başlık ile).
struct DiscoveredBackend: Identifiable {
    let name: String      // ör. "Ollama", "LM Studio"
    let baseURL: String   // ör. "http://localhost:11434/v1"
    let models: [String]  // /v1/models -> data[].id
    var id: String { baseURL }
}

/// Bilinen yerel OpenAI-uyumlu sunucuları tarar ve modelleri **türlerine göre**
/// (Ollama, LM Studio, …) listeler. Her biri `/v1/models` ucuyla yoklanır.
enum ModelDiscovery {
    /// (görünen ad, base_url). Yaygın yerel sunucuların varsayılan portları.
    static let known: [(name: String, baseURL: String)] = [
        ("Ollama", "http://localhost:11434/v1"),
        ("LM Studio", "http://localhost:1234/v1"),
        ("llama.cpp", "http://localhost:8080/v1"),
        ("Jan", "http://localhost:1337/v1"),
    ]

    /// Tüm bilinen sunucuları paralel yoklar; erişilebilen + modeli olanları döner
    /// (known sırasını koruyarak).
    static func discover() async -> [DiscoveredBackend] {
        let found = await withTaskGroup(of: DiscoveredBackend?.self) { group -> [DiscoveredBackend] in
            for backend in known {
                group.addTask { await probe(name: backend.name, baseURL: backend.baseURL) }
            }
            var result: [DiscoveredBackend] = []
            for await item in group {
                if let item { result.append(item) }
            }
            return result
        }
        return known.compactMap { entry in found.first { $0.baseURL == entry.baseURL } }
    }

    private static func probe(name: String, baseURL: String) async -> DiscoveredBackend? {
        guard let url = URL(string: baseURL + "/models") else { return nil }
        var request = URLRequest(url: url)
        request.timeoutInterval = 1.5  // yerel; kapalı sunucu anında reddedilir
        guard
            let (data, response) = try? await URLSession.shared.data(for: request),
            (response as? HTTPURLResponse)?.statusCode == 200,
            let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let array = object["data"] as? [[String: Any]]
        else { return nil }

        let models = array.compactMap { $0["id"] as? String }.sorted()
        return models.isEmpty ? nil : DiscoveredBackend(name: name, baseURL: baseURL, models: models)
    }
}

File: windows/Turkify/ModelDiscovery.cs
using System.IO;
using System.Net.Http;
using System.Text.Json;

namespace Turkify;

/// Keşfedilen bir yerel LLM sunucusu ve modelleri (tür/başlık ile).
public sealed record DiscoveredBackend(string Name, string BaseURL, IReadOnlyList<string> Models);

/// Bilinen yerel OpenAI-uyumlu sunucuları tarar ve modelleri **türlerine göre**
/// (Ollama, LM Studio, …) listeler. Her biri <c>/v1/models</c> ucuyla yoklanır.
/// (macOS ModelDiscovery.swift'in karşılığı.)
public static class ModelDiscovery
{
    /// (görünen ad, base_url). Yaygın yerel sunucuların varsayılan portları.
    private static readonly (string Name, string BaseURL)[] Known =
    [
        ("Ollama", "http://localhost:11434/v1"),
        ("LM Studio", "http://localhost:1234/v1"),
        ("llama.cpp", "http://localhost:8080/v1"),
        ("Jan", "http://localhost:1337/v1"),
    ];

    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(1.5) };

    /// Tüm bilinen sunucuları paralel yoklar; erişilebilen + modeli olanları döner
    /// (known sırasını koruyarak).
    public static async Task<IReadOnlyList<DiscoveredBackend>> DiscoverAsync()
    {
        IEnumerable<Task<DiscoveredBackend?>> probes =
            Known.Select(entry => ProbeAsync(entry.Name, entry.BaseURL));
        DiscoveredBackend?[] results = await Task.WhenAll(probes).ConfigureAwait(false);
        return results.Where(b => b is not null).Cast<DiscoveredBackend>().ToList();
    }

    private static async Task<DiscoveredBackend?> ProbeAsync(string name, string baseURL)
    {
        try
        {
            using HttpResponseMessage response = await Http.GetAsync(baseURL + "/models").ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                return null;
            }

            await using Stream stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
            using JsonDocument document = await JsonDocument.ParseAsync(stream).ConfigureAwait(false);
            if (!document.RootElement.TryGetProperty("data", out JsonElement data) ||
                data.ValueKind != JsonValueKind.Array)
            {
                return null;
            }

            var models = data.EnumerateArray()
                .Select(item => item.TryGetProperty("id", out JsonElement id) ? id.GetString() : null)
                .Where(id => !string.IsNullOrEmpty(id))
                .Cast<string>()
                .OrderBy(id => id, StringComparer.OrdinalIgnoreCase)
                .ToList();

            return models.Count == 0 ? null : new DiscoveredBackend(name, baseURL, models);
        }
        catch (Exception)
        {
            // Kapalı/erişilemeyen sunucu: sessizce atla (keşif best-effort).
            return null;
        }
    }
}
